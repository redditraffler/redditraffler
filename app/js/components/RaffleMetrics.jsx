import React from "react";
import useAsync from "react-use/lib/useAsync";
import Skeleton from "react-loading-skeleton";
import styled from "styled-components";
import Columns from "react-bulma-components/lib/components/columns";
import Box from "react-bulma-components/lib/components/box";
import Container from "react-bulma-components/lib/components/container";
import Heading from "react-bulma-components/lib/components/heading";
import Tile from "react-bulma-components/lib/components/tile";

import { getRaffleMetrics } from "@js/api";
import { colors } from "@js/theme";

const StatHeading = styled(Heading)`
  color: ${colors.reddit};
  text-align: center;
`;

const EqualHeightBox = styled(Box)`
  display: flex;
  flex-direction: column;
  height: 100%;
  border-radius: 10px;
`;

const RaffleMetrics = () => {
  const { loading, error, value: metrics } = useAsync(getRaffleMetrics);

  if (error) {
    return (
      <Container>
        <Columns centered>
          <Columns.Column size="one-quarter">
            <Box size="one-quarter">Couldn&apos;t fetch raffle metrics :(</Box>
          </Columns.Column>
        </Columns>
      </Container>
    );
  }

  const metricsForDisplay = [
    {
      displayValue: metrics?.num_total_subreddits?.toLocaleString(),
      label: "communities served",
    },
    {
      displayValue: metrics?.num_total_verified_raffles?.toLocaleString(),
      label: "raffles created",
    },
    {
      displayValue: metrics?.num_total_winners?.toLocaleString(),
      label: "lucky winners",
    },
  ];

  return (
    <Container>
      <Columns centered breakpoint="desktop">
        <Columns.Column size={6}>
          <EqualHeightBox>
            <Tile kind="parent" style={{ alignItems: "center" }}>
              {loading ? (
                <Tile kind="child">
                  <Skeleton count={2} height={30} />
                </Tile>
              ) : (
                <React.Fragment>
                  {metricsForDisplay.map(({ displayValue, label }) => {
                    return (
                      <Tile kind="child" key={label}>
                        <StatHeading>{displayValue}</StatHeading>
                        <Heading subtitle style={{ textAlign: "center" }}>
                          {label}
                        </Heading>
                      </Tile>
                    );
                  })}
                </React.Fragment>
              )}
            </Tile>
          </EqualHeightBox>
        </Columns.Column>
      </Columns>
    </Container>
  );
};

export default RaffleMetrics;
