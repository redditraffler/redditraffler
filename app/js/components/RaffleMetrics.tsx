import React from "react";
import useAsync from "react-use/lib/useAsync";
import Skeleton from "react-loading-skeleton";
import styled from "styled-components";
import Columns from "react-bulma-components/lib/components/columns";
import Box from "react-bulma-components/lib/components/box";
import Container from "react-bulma-components/lib/components/container";
import Heading from "react-bulma-components/lib/components/heading";
import Tile from "react-bulma-components/lib/components/tile";

import { Endpoint, getFromApi } from "@js/api";
import { colors } from "@js/theme";
import { RaffleMetricsAPIResponse } from "@js/types";

const StatHeading = styled(Heading)`
  color: ${colors.reddit};
  text-align: center;
`;

const MetricsBox = styled(Box)`
  border-radius: 10px;
`;

const RaffleMetrics: React.FC = () => {
  const { loading, error, value: metrics } = useAsync(() =>
    getFromApi<RaffleMetricsAPIResponse>(Endpoint.getRaffleMetrics)
  );

  if (error) {
    // @ts-ignore
    return (
      <Container>
        <Columns centered>
          <Columns.Column size="one-quarter">
            <Box size="one-quarter">Couldn&apos;t fetch raffle metrics :(</Box>
          </Columns.Column>
        </Columns>
      </Container>
    );
  }

  const metricsForDisplay = [
    {
      displayValue: metrics?.num_total_subreddits?.toLocaleString(),
      label: "communities served",
    },
    {
      displayValue: metrics?.num_total_verified_raffles?.toLocaleString(),
      label: "raffles created",
    },
    {
      displayValue: metrics?.num_total_winners?.toLocaleString(),
      label: "lucky winners",
    },
  ];

  // @ts-ignore
  return (
    <Container>
      <Columns centered>
        <Columns.Column size={6}>
          <MetricsBox>
            <Tile kind="parent" style={{ alignItems: "center" }}>
              {loading ? (
                <Tile kind="child">
                  <Skeleton count={2} height={30} />
                </Tile>
              ) : (
                <React.Fragment>
                  {metricsForDisplay.map(({ displayValue, label }) => (
                    <Tile kind="child" key={label}>
                      <StatHeading>{displayValue}</StatHeading>
                      <Heading subtitle style={{ textAlign: "center" }}>
                        {label}
                      </Heading>
                    </Tile>
                  ))}
                </React.Fragment>
              )}
            </Tile>
          </MetricsBox>
        </Columns.Column>
      </Columns>
    </Container>
  );
};

export default RaffleMetrics;
