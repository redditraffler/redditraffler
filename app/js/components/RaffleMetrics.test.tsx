import React from "react";
import axios from "redaxios";
import { render, screen } from "@testing-library/react";

import RaffleMetrics from "./RaffleMetrics";

jest.mock("redaxios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe("when the API call is in progress", () => {
  beforeEach(() => {
    mockedAxios.get.mockImplementation((async () => {
      await new Promise((resolve) => setTimeout(99999, resolve));
    }) as any); // eslint-disable-line @typescript-eslint/no-explicit-any
  });

  it("returns the skeleton", async () => {
    const { container } = render(<RaffleMetrics />);

    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("when the API call fails", () => {
  beforeEach(() => {
    mockedAxios.get.mockImplementation(async () => {
      throw new Error("oh boy");
    });
  });

  it("returns the error message", async () => {
    const { container } = render(<RaffleMetrics />);
    await screen.findByText(/fetch raffle metrics/);

    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("when the API call is successful", () => {
  beforeEach(() => {
    mockedAxios.get.mockResolvedValue({
      data: {
        num_total_verified_raffles: 123,
        num_total_winners: 456,
        num_total_subreddits: 1234,
        top_recent_subreddits: [],
      },
    } as any); // eslint-disable-line @typescript-eslint/no-explicit-any
  });

  it("returns the metrics container", async () => {
    const { container } = render(<RaffleMetrics />);
    await screen.findByText(/communities served/);

    expect(container.firstChild).toMatchSnapshot();
  });
});
