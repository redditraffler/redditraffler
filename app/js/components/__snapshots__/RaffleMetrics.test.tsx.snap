// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the API call fails returns the error message 1`] = `
<div
  class="container"
>
  <div
    class="columns is-multiline is-centered"
  >
    <div
      class="column is-one-quarter"
    >
      <div
        class="box"
      >
        Couldn't fetch raffle metrics :(
      </div>
    </div>
  </div>
</div>
`;

exports[`when the API call is in progress returns the skeleton 1`] = `
.c0 {
  border-radius: 10px;
}

<div
  class="container"
>
  <div
    class="columns is-multiline is-centered"
  >
    <div
      class="column is-6"
    >
      <div
        class="box c0"
      >
        <div
          class="tile is-parent"
          style="align-items: center;"
        >
          <div
            class="tile is-child"
          >
            <span>
              <span
                class="react-loading-skeleton css-1vmnjpn-skeletonStyles-Skeleton"
                style="height: 30px;"
              >
                ‌
              </span>
              <span
                class="react-loading-skeleton css-1vmnjpn-skeletonStyles-Skeleton"
                style="height: 30px;"
              >
                ‌
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`when the API call is successful returns the metrics container 1`] = `
.c1 {
  color: #ff7644;
  text-align: center;
}

.c0 {
  border-radius: 10px;
}

<div
  class="container"
>
  <div
    class="columns is-multiline is-centered"
  >
    <div
      class="column is-6"
    >
      <div
        class="box c0"
      >
        <div
          class="tile is-parent"
          style="align-items: center;"
        >
          <div
            class="tile is-child"
          >
            <h1
              class="c1 title"
            >
              1,234
            </h1>
            <h1
              class="subtitle"
              style="text-align: center;"
            >
              communities served
            </h1>
          </div>
          <div
            class="tile is-child"
          >
            <h1
              class="c1 title"
            >
              123
            </h1>
            <h1
              class="subtitle"
              style="text-align: center;"
            >
              raffles created
            </h1>
          </div>
          <div
            class="tile is-child"
          >
            <h1
              class="c1 title"
            >
              456
            </h1>
            <h1
              class="subtitle"
              style="text-align: center;"
            >
              lucky winners
            </h1>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
